Allow Basic authentication in browser

nivertius opened the pull request:
Fix for #88

This change allows setting new command line option `-basicauth` that allows sending files to authfile-enabled servers through browser.
When this option is enabled, upload and paste page is shown even when `-authfile` option is present. After either uploading a file, or pasting text, an standard browser authentication window is shown. Username is ignored, and password is taken as Api-Key. If the password match entry in authfile, upload or paste is allowed. If not, it fails with Unauthorized message.
Because browsers typically keep `Authorization` header, successive uploads don't require another password entry.

Unfortunately, there's no indication for user that password is the field for api key.

Please note that I do not have any experience in Go, and this was mostly done by reverse engineering already existing code. I tried to keep the changes as small as possible and everything should work as before if new command line options is not passed. I will be more than happy for guidance or tips how to improve the solution, or just straight up amends, if this is not adequate.

andreimarcu commented:
Thanks for the PR. I tried it and noticed the following:
- The Upload, Paste links are missing from navigation 
- Uploads will just fail if you're uploading via the Upload page since the javascript won't perform the auth request. I only got it to work after trying to paste, getting the basic auth prompt and then uploading through the Upload page. 

For the latter, one way would be to have the uploader ask for the key and then use it in its requests

nivertius commented:
* I added a check to show Upload and Paste links on main page, if basic auth is enabled.
* I added handling basic auth on `upload/` page, if key is not provided from `FormValue` it is read from the authorization header.

I'm sorry, but I don't follow your second bullet point. In my tests both on Firefox and Brave I confirmed following scenario to occur:
* User is not logged (i.e. doesn't send `Authorization` header)
* User goes to index, is shown dropzone.
* User tries one of the following: pastes a file from clipboard, ex. image or clicks on the dropzone and selects file from local machine.
* Javascript processes the file and tries to upload it.
* Server sees that there isn't a `Linx-Api-Key` or `Authorization` header and responds with `401 Unauthorized` status
* Browser receives the response and shows native window with two fields.
* User types anything into username, and valid key from authfile into password and submits the form.
* Javascript **retries** request, browser adds `Authorization` header.
* Servers verifies the that there still isn't `Linx-Api-Key`, but there's a `Authorization` header now and allows upload.
* User is show uploaded entry.

Successive requests for upload skip the login window, because browser adds the header to the first request.

I did not dive into `dropzone.js` internals to confirm that request is repeated, but this is what I observe.

If this isn't what your second point was about, could you clarify if please?

I could also maybe add "Authorize" link before "Upload | Paste | API" links, but this would be somewhat complicated. Let's say this link would go to `auth/` path, which would be handled by simple handler that checks for basic auth header and redirects to index. Then, the "Authorize" link should disappear, probably replaced by "Authorized" message. This in turn needs to be checked for every request. There's also problem with what to do if user authorized by sending file, with scenario above. I don't think this adds much for how it's complicated.

If above scenario works as I observe it, this makes authorization pretty simple: if you try to upload, you need to authorize, and if you uploaded, you are authorized. But of course I'm open for suggestion.

andreimarcu commented:
Hey, thanks for working on this. 
I confirmed Chrome performs as you describe, however on Firefox the uploader will not prompt for the basic auth form. 
So it does look like we're gonna have to dig in the dropzone code to check for a 401 return and prompt the user for the key, then retrying the request including the basic auth header. 
I can't work on this right now so feel free to try your hand at it, if not I'll come back to it later

andreimarcu commented:
This is weird, it seems that on Firefox, small files will prompt for basic auth while larger ones get an empty response in the uploader (which is why it doesn't prompt for basic auth on those).

I can't investigate further right now but if you want you can try to figure out why dropzone does that (seems like the server is returning 401 correctly even though the uploader gets an empty response)

nivertius commented:
I can confirm that trying to upload larger files makes server return "status 0", i.e. empty response.

After some investigation this is because net/http server rejects unread requests larger than 262144 bytes. See https://github.com/golang/go/blob/master/src/net/http/server.go#L1317 .
This happens because after receiving the headers, handler for upload is executed, which in case of no authentication header doesn't even read the contents, just writes 401 and returns. So the whole request is unprocessed, when server tries to actually write the response. For the reasons above (preventing DoS), server doesn't write anything, just closes connection.

This means that my solution doesn't work in general and it needs to be reworked. There should be check before user tries to upload file if he actually can. 

I see two solutions:
* If basic auth is enabled, send a HEAD request first, to check if authentication is required.
* Don't show either dropzone or Paste link until user authenticates - for example, as I wrote before, with "Authenticate" link instead of "Upload | Paste"

I think the second one is more robust, but I don't know if I have enough knowledge about golang to do it ;-)

andreimarcu commented:
Great catch @nivertius! 

I'm leaning towards the following: 
- If your linx instance has -basicauth enabled, include a data attribute in the uploader <form> (like the current data-maxsize) that indicates basicauth is enabled
- in the dropzone config, before an upload fires, check if the data attribute is there. If so, check if the dropzone headers config includes the Authorization header. If not, prompt the user for the key, and then add it. 
- From there uploads will have the auth header and will succeed

I could take care of this if you want in the next few days

nivertius commented:
I did more research: This problem should also happen with regular (the `Linx-Api-Key` header) authentication - if user sends `PUT` to upload with 'large' file and no header, server should close connection after sending status, without reading the whole request, as it is seen here. And this actually happens!

But not when you send with cURL. It actually sends 'empty' `PUT` request with `Expect: 100-continue` header, and then, if 100 happens, sends the rest of the request. This header is exactly what we need here: 
* [Expect header on MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Expect)
* [RFC2616 (HTTP/1.1), section 8.2.3](https://www.w3.org/Protocols/rfc2616/rfc2616-sec8.html#sec8.2.3)

What we could do is force Dropzone, or just browser to send headers with the expect header first. But unfortunately, as far as I can tell, no browsers support this, and `Expect` header is on [Forbidden header list](https://developer.mozilla.org/en-US/docs/Glossary/Forbidden_header_name) and cannot be set on XHR.

nivertius commented:
After even more research, and coming to conclusion that you can't force browser to do `Expect: 100-continue` I did it manually:
* Added a separate handler for `auth/` path that only authenticates.
* If basic authentication is needed, automatic upload queue processing on client is disabled. Only when user selects a file, then single `HEAD` request to above handler is sent. If server returns 2xx or 3xx, then queue is manually processed and automatic processing is enabled. If not, file is marked as canceled.

This means that if user does is not authorized, file upload will not happen, and this is what we need.

andreimarcu commented:
I'll give it a look hopefully in the next few days. Thanks for your patience

andreimarcu commented:
Thank you for working on this!

nivertius commented:
My pleasure. Thanks for merging.

Could you tell me when this could be released?

andreimarcu commented:
Around now!
